dz2

PretenderW opened the pull request:
Dz1 - изменение котролеров 
Dz2 - миграции и сиды

ematsan committed:
dz2

NTlegend commented on start/routes.js:
```diff
@@ -15,3 +15,11 @@
+Route.resource('products', 'ProductController');
+Route.resource('types', 'TypeController');
+Route.resource('atributes', 'AtributeController');
+Route.resource('productatributes', 'ProductAtributeController');
```
это не самостоятельная сущьность, а связь продуктов и аттрибутов. т.к. ключевыми в нашем приложении выступают продукты - то эту связь логичнее сохранять вместе с продуктами

PretenderW replied:
Route.resource('products/productatributes', 'ProductAtributeController').apiOnly();

PretenderW replied:
так устроит?

NTlegend replied:
при покупке телефона в магазине ты же не платишь отдельно за телефон и отдельно за диагональ, так а почему тут ты хочешь получить отдельно телефон и отдельно характеристики телефона?

PretenderW replied:
Все понял: убрать атрибуты продукта, а всю обработку осуществлять в продукте.

NTlegend commented on start/routes.js:
```diff
@@ -15,3 +15,11 @@
+
+Route.resource('products', 'ProductController');
+Route.resource('types', 'TypeController');
+Route.resource('atributes', 'AtributeController');
```
русурсный контроллер подразумевает что у тебя будут методы для рендера форм.. чтобы откинуть эти роуты добавь [apiOnly()](https://adonisjs.com/docs/4.1/routing#_apionly)

PretenderW replied:
Route.resource('products', 'ProductController').apiOnly();
Route.resource('types', 'TypeController').apiOnly();
Route.resource('atributes', 'AtributeController').apiOnly();
Route.resource('products/productatributes', 'ProductAtributeController').apiOnly();

NTlegend commented on database/migrations/1546262372089_product_atribute_schema.js:
```diff
@@ -0,0 +1,35 @@
+  up() {
+    this.create('product_atributes', table => {
+      table.increments();
+      table.varchar('value', 255).notNullable();
```
значение 255 - по умолчанию, его можно не писать

PretenderW replied:
Зависимость от значений по умолчанию не всегда обоснована. В данном случае у меня стабильный размер поля независимый от значений по умолчанию.
Да и запоминать все значения по умолчанию в различных СУБД это лишняя трата времени и сил.

NTlegend commented on database/seeds/01UserSeeder.js:
```diff
@@ -0,0 +1,29 @@
+*/
+
+/** @type {import('@adonisjs/lucid/src/Factory')} */
+// const Factory = use('Factory');
```
в репозитории не должно быть закоментированого кода

PretenderW replied:
убрал все автоматические комментарии

NTlegend commented on database/seeds/03ProductSeeder.js:
```diff
@@ -0,0 +1,30 @@
+    await Product.query().delete();
+
+    const products = [
+      { name: 'asus', user_id: 1, type_id: 1 },
```
при повтором запуске сидов у тебя же будет другие ID-шки для юзеров и типов, выбирай их с БД, чтобы не хардкодить

PretenderW replied:
const types = await Type.pair('name', 'id');
    const users = await User.pair('name', 'id');

    const products = [
      { name: 'asus', user_id: users.Ivanov, type_id: types.nout },
      { name: 'lenovo', user_id: users.Petrov, type_id: types.phone },
      { name: 'audi', user_id: users.Petrov, type_id: types.car },
      { name: 'xiaomi', user_id: users.Sidorov, type_id: types.phone }
   ];

NTlegend commented on database/seeds/05ProductAtributeSeeder.js:
```diff
@@ -0,0 +1,32 @@
+    const productatributes = [
+      { product_id: 1, atribute_id: 1, value: '15"' },
+      { product_id: 2, atribute_id: 2, value: '4.5"' },
+      { product_id: 2, atribute_id: 4, value: 'S820' },
```
для тестовых данных у всех продуктов должны быть заполнены все аттрибуты в соответствии с типами

PretenderW replied:
const datalist = await Database.select('products.id as prod_id', 'atributes.id as atr_id')
      .from('products')
      .leftJoin('atributes', 'products.type_id', 'atributes.type_id');
    let productatributes = [];
    for (let i in datalist)
      productatributes.push({ product_id: datalist[i].prod_id, atribute_id: datalist[i].atr_id, value: i });

PretenderW replied:
только теперь данные у атрибута произвольные

NTlegend commented on app/Models/Product.js:
```diff
@@ -0,0 +1,14 @@
+    return null;
+  }
+
+  static get createdAtColumn() {
```
Из задания:
```
 - не авторизированый пользователь:
    - просматривает списки товаров с сортировкой по дате добавления
```

PretenderW replied:
А зачем сортировку реализовывать в модели?
Я ее реализую в запросе используя orderBy

NTlegend replied:
как ты отсортируешь, если у тебя нет этого поля

PretenderW replied:
Есть это поле, и заполняется авоматически table.timestamp('created_at').defaultTo(this.fn.now());

ematsan committed:
correct remark
 Changes to be committed:
	modified:   app/Controllers/Http/AtributeController.js
	modified:   app/Controllers/Http/ProductAtributeController.js
	modified:   app/Controllers/Http/ProductController.js
	modified:   app/Controllers/Http/TypeController.js
	modified:   app/Models/Atribute.js
	modified:   app/Models/Product.js
	modified:   app/Models/ProductAtribute.js
	modified:   app/Models/Type.js
	modified:   database/migrations/1546262320748_type_schema.js
	modified:   database/migrations/1546262344878_atribute_schema.js
	modified:   database/migrations/1546262354116_product_schema.js
	modified:   database/migrations/1546262372089_product_atribute_schema.js
	modified:   database/seeds/01UserSeeder.js
	modified:   database/seeds/02TypeSeeder.js
	modified:   database/seeds/03ProductSeeder.js
	modified:   database/seeds/04AtributeSeeder.js
	modified:   database/seeds/05ProductAtributeSeeder.js
	modified:   start/routes.js

NTlegend commented on start/routes.js:
```diff
@@ -1,25 +1,11 @@
+Route.resource('products', 'ProductController').apiOnly();
+Route.resource('types', 'TypeController').apiOnly();
+Route.resource('atributes', 'AtributeController').apiOnly();
+Route.resource('products/productatributes', 'ProductAtributeController').apiOnly();
```
это часть информации о продукте, боле логично менеджить ее вместе с продуктом.

PretenderW replied:
Все понял. Просто пока запросов нету это замечание не является явным. Как запросы начнут писаться все сразу же проясниться.

NTlegend commented on database/seeds/05ProductAtributeSeeder.js:
```diff
@@ -1,30 +1,18 @@
+    let productatributes = [];
+    for (let i in datalist)
+      productatributes.push({ product_id: datalist[i].prod_id, atribute_id: datalist[i].atr_id, value: i });
+    //    console.log(productatributes);
```
`console.log` лишний

PretenderW replied:
Это я проверял/тестировал заполнения данных. Удалил

NTlegend commented on database/seeds/05ProductAtributeSeeder.js:
```diff
@@ -1,30 +1,18 @@
+    for (let i in datalist)
+      productatributes.push({ product_id: datalist[i].prod_id, atribute_id: datalist[i].atr_id, value: i });
+    //    console.log(productatributes);
     await ProductAtribute.createMany(productatributes);
```
нужно паравильно настроить релейшены в моделе продуктов и сохранять через эти релейшены

PretenderW replied:
Релейшены добавил.
Но продукты с атрибутами связаны через тип и не совсем понятно как это сделать через релейшены.

NTlegend commented on app/Models/Product.js:
```diff
@@ -1,4 +1,3 @@
-/** @type {typeof import('@adonisjs/lucid/src/Lucid/Model')} */
 const Model = use('Model');
 
 class Product extends Model {
```
из задания:
```
  -  не авторизированый пользователь:
      - просматривает списки товаров с сортировкой по дате добавления
```
у тебя это возможности нет, т.к. нет у продукта даты добавления

ematsan committed:
delete productatribute controller
 Changes to be committed:
	deleted:    app/Controllers/Http/ProductAtributeController.js
	modified:   database/seeds/05ProductAtributeSeeder.js
	modified:   start/routes.js

ematsan committed:
add relations
 Changes to be committed:
	modified:   app/Models/Atribute.js
	modified:   app/Models/Product.js
	modified:   app/Models/ProductAtribute.js
	modified:   app/Models/Type.js
	modified:   app/Models/User.js
